### PR TITLE
PHP 7.3: supporting native decoration-free logs

### PR DESCRIPTION
**Pull request by bryanlatten**
- [x] Depends on https://github.com/behance/docker-nginx/pull/67
- [x] workaround for goss YAML deep merge on `file` test directive
- Forks startup behavior based on 7.3 + new `ini` directive or not
- Improves shutdown behavior. 
- Added inline startup check for fpm syntax. Prevents bad env variable inputs from producing invalid config

### CODE REVIEW

File: Dockerfile-73 [hunk 1]
```diff
@@ -133,8 +135,16 @@ RUN cp /etc/php/7.0/mods-available/* $CONF_PHPMODS && \
     phpenmod overrides && \
     # Enable NewRelic via Ubuntu symlinks, but disable in file. Cross-variant startup script uncomments with env vars.
     phpenmod newrelic && \
-    # - Run standard set of tweaks to ensure runs performant, reliably, and consistent between variants
-    /bin/bash -e /prep-php.sh
+    # Run standard set of tweaks to ensure runs performant, reliably, and consistent between variants
+    /bin/bash -e /prep-php.sh && \
+    # Add new 7.3+ conf
+    sed -i "s/;decorate_workers_output.*/decorate_workers_output = no/" $CONF_FPMPOOL && \
```
Comment (mlebrun) [inline Dockerfile-73]: is `decorate_workers_output = no` backwards compatible? or is it necessary _to be_ backwards compatible?

File: container/root/etc/services-available/php-fpm/.command [hunk 2]
```diff
@@ -3,4 +3,13 @@
 # Ensures that a failure of php-fpm doesn't return with a 0, even if the right-hand side does
 set -o pipefail
 
-exec php-fpm -F -O 2>&1 | sed -u 's,.*: \"\(.*\)$,\1,'| sed -u 's,\"$,,'
+# IMPORTANT: PHP 7.3+ optionally allows undecorated stdout/stderr, removing pipe magic requirement
+UPGRADE_COMMAND=`grep "decorate_workers_output = no" $CONF_FPMPOOL`
```
Comment (mlebrun) [inline container/root/etc/services-available/php-fpm/.command:7]: this is probably what makes it backwards compatible?

File: container/root/etc/services-available/php-fpm/finish [hunk 3]
```diff
@@ -4,4 +4,7 @@
 if { s6-test ${1} -ne 0 }
 if { s6-test ${1} -ne 256 }
 
-s6-svscanctl -t /var/run/s6/services
+# Suppress superfluous time-of-check-time-of-use (TOCTOU) message
```
Comment (mlebrun) [inline container/root/etc/services-available/php-fpm/finish:7]: is `time-of-check-time-of-use` or is it supposed to be just `check-time-of-use`?

File: container/root/etc/services-available/php-fpm/run [hunk 4]
```diff
@@ -8,22 +8,11 @@ trap
...
-      {
-        echo [sigterm-phpfpm] waiting for nginx to exit
-      }
-      wait { $NGINXPID } import -u ! kill -s SIGQUIT ${!}
-    }
-    foreground
-    {
-      import -u ! kill -s SIGQUIT ${!}
-    }
+    echo [run php-fpm] TERM received
```
Comment (mlebrun) [inline container/root/etc/services-available/php-fpm/run:11]: these are just echoing now and not actually doing anything?

File: Dockerfile-73 (hunk 1 above)
Comment (bryanlatten) [reply Dockerfile-73]: no, it only exists in 7.3+

File: container/root/etc/services-available/php-fpm/finish (hunk 3 above)
Comment (bryanlatten) [reply container/root/etc/services-available/php-fpm/finish:7]: its specifically TOCTOU

File: container/root/etc/services-available/php-fpm/run (hunk 4 above)
Comment (bryanlatten) [reply container/root/etc/services-available/php-fpm/run:11]: bingo. we're suppressing these signals: 
https://github.com/behance/docker-base#implementing-graceful-shutdown-from-admin-initiated-container-termination

File: container/root/etc/services-available/php-fpm/.command (hunk 2 above)
Comment (bryanlatten) [reply container/root/etc/services-available/php-fpm/.command:7]: exactly. the `ini` directive only exists where valid (7.3 and on)

**Comment by DFoxinator** [comment]
Awesome, very excited to try the logging changes.

**Review by DFoxinator — APPROVED** [review]

**Review by mlebrun — APPROVED** [review]

File: Dockerfile-73 [hunk 5]
```diff
@@ -138,10 +136,7 @@ RUN cp /etc/php/7.0/mods-available/* $CONF_PHPMODS && \
     # Run standard set of tweaks to ensure runs performant, reliably, and consistent between variants
     /bin/bash -e /prep-php.sh && \
     # Add new 7.3+ conf
-    sed -i "s/;decorate_workers_output.*/decorate_workers_output = no/" $CONF_FPMPOOL && \
-    sed -i "s/;log_limit.*/log_limit = \${PHP_FPM_LOG_LIMIT}/" $CONF_PHPFPM && \
```
Comment (bryanlatten) [inline Dockerfile-73]: environment variables not valid in `php-fpm.conf` file, moved to startup (run.d) script replacement.

**Review by mlebrun — APPROVED** [review]

**Review by DFoxinator — APPROVED** [review]

**Pull request merged**